packages/utils/src/rateLimiterGRCA.ts

```diff
@@ -0,0 +1,124 @@
+type MiliSeconds = number;
+
+export interface RateLimiterQuota {
+  /** How often are `max_tokens` fully replenished. */
+  replenishAllEvery: MiliSeconds;
+  /** Token limit. This translates on how large can an instantaneous batch of tokens be. */
+  maxTokens: number;
+}
+
+/**
+ * Generic Cell Rate Algorithm is a leaky bucket-type scheduling algorithm.
+ *
+ * Most rate-limit implementations are either time-bucket or leaky-bucket based. The time-bucket requires the storage
+ * of two values and does not enforce a rate, while the leaky-bucket approach requires a separate process to
+ * continually refill the bucket. GCRA only storing a value (the TAT) while still being simple. GCRA may be rarely
+ * used because of its perceived complexity.
+ *
+ * GCRA aims to limit requests to `R = L/P`, where this implementation sets `L = 1` for simplicity. The target rate
+ * then is `R = 1/P` so request separated by at least `P` are not limited. Define the Theoretical Arrival Time (TAT)
+ * of the next request to be equal
+ */
+export class RateLimiterGRCA<Key> {
+  /** Time when the bucket will be full for each peer. TAT (theoretical arrival time) from GCRA */
+  private readonly tatPerKey = new Map<Key, MiliSeconds>();
+  private readonly startTimeMs = Date.now();
+
+  constructor(
+    /** After how long is the bucket considered full via replenishing 1T every `t`. */
+    private readonly msPerBucket: MiliSeconds,
+    /** How often is 1 token replenished */
+    private readonly msPerToken: MiliSeconds
+  ) {}
+
+  static fromQuota<Key>(quota: RateLimiterQuota): RateLimiterGRCA<Key> {
+    if (quota.maxTokens === 0) {
+      throw Error("Max number of tokens should be positive");
+    }
+    const msPerBucket = quota.replenishAllEvery;
+    if (msPerBucket === 0) {
+      throw Error("Replenish time must be positive");
+    }
+    const msPerToken = msPerBucket / quota.maxTokens;
+    return new RateLimiterGRCA(msPerBucket, msPerToken);
+  }
+
+  currentBucketRemainingTime(key: Key, msSinceStart?: number): number {
```

Comment:
@dapplion I used this rate limiter implementation from discv5. I believe this additional parameters `msSinceStart` was added for the testing purpose only and don't have a real use case. Should we remove it?

Reply:
You should copy all the unit tests to ensure this copied source code doesn't break

Reply:
Hmm un-sure would have to re-review the logic